### PR TITLE
Redesign `ReadStatMeta` and add `ReadStatColMeta` for DataAPI.jl v1.13

### PR DESCRIPTION
**Pull request by nalimilan**
Use the `metadata` function that is going to be added to DataAPI (https://github.com/JuliaData/DataFrames.jl/pull/3055) to expose the contents of `ReadStatMeta`. This will allow e.g. `DataFrame` to import the metadata.

The implementation is not super efficient as a new `Dict` is allocated for each column. If this turns out to be a problem it would be possible to create a special `AbstractDict` type that would expose the data without allocating, but this may be overkill. Also, if users mutate the returned metadata `Dict`s, the changes will have no effect on `ReadStatMeta`, except if they mutate the `Dict`s holding value labels. We could copy these `Dict`s for safety, as mutating them for one variable will affect all other variables that use the same value labels.

Cc: @bkamins

### CODE REVIEW

**Comment by bkamins** [comment]
@nalimilan Regarding performance, as you have mentioned this several times. Do you think anyone will notice such delay if one has 10,000 column data frame to work with:
```
julia> @benchmark Dict(i => Dict{String, Any}() for i in 1:10_000)
BenchmarkTools.Trial: 4141 samples with 1 evaluation.
 Range (min … max):  921.400 μs …   7.387 ms  ┊ GC (min … max):  0.00% … 69.63%
 Time  (median):     979.800 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.204 ms ± 610.899 μs  ┊ GC (mean ± σ):  17.04% ± 20.13%

  ▇█▆▄▂▂                    ▂                      ▁            ▁
  ████████▇▅▆▅▅▄▅▃▄▃▅▁▁▃▁▁▃███▇▇██▇▇▅▅▅▅▅▃▅▃▃▅▃▆▇█████▇▆▆▅▅▆▅▅▆ █
  921 μs        Histogram: log(frequency) by time       3.54 ms <

 Memory estimate: 5.15 MiB, allocs estimate: 40007.
```

**Comment by nalimilan** [comment]
An overhead of about 1ms sounds reasonable (and 10,000 columns is relatively large). But it would be worth benchmarking the PR to check this.

**Comment by junyuan-chen** [comment]
Thanks for the great work on metadata! With the v1.12 release of `DataAPI.jl` and v1.4 release of `DataFrames.jl`, now it is the time to adopt the new interface.

Since the changes made here are already outdated, I am starting to rewrite the part on metadata. There is no good reason to keep the old design with `ReadStatMeta`. So, I will simply make some breaking changes that will lead to v0.2.

**Comment by bkamins** [comment]
@junyuan-chen thank you! Just please note that in https://github.com/JuliaData/DataAPI.jl/pull/56 some minimal extensions to the metadata API are added (nothing breaking, just small changes that make using metadata more convenient; they should be merged probably in the coming week). Thank you!

**Comment by junyuan-chen** [comment]
@bkamins Thanks for mentioning that! I will keep that in mind.

File: Project.toml [hunk 1]
```diff
@@ -13,7 +13,7 @@ Tables = "bd369af6-aec1-5ad0-b16a-f7cc5008161c"
 
 [compat]
 CategoricalArrays = "0.10"
-DataAPI = "1.6"
+DataAPI = "1.11"
```
Comment (bkamins) [inline Project.toml:16]: I would prepare this PR against DataAPI.jl 1.13 (it will be released this week)

File: src/table.jl [hunk 2]
```diff
@@ -123,6 +123,53 @@ Retrieve the metadata parsed from a data file.
 """
 getmeta(tb::ReadStatTable) = getfield(tb, :meta)
 
+hasmetadata(tb::ReadStatTable) = getmeta(tb) !== nothing
```
Comment (bkamins) [inline src/table.jl:126]: it should be `metadatasupport` and it takes type as an argument.

File: src/table.jl [hunk 3]
```diff
@@ -123,6 +123,53 @@ Retrieve the metadata parsed from a data file.
...
+hasmetadata(tb::ReadStatTable, col::Symbol) = getmeta(tb) !== nothing && haskey(tb, col)
+
+function metadata(tb::ReadStatTable)
```
Comment (bkamins) [inline src/table.jl:129]: also `style` kwarg should be supported.

File: src/table.jl [hunk 4]
```diff
@@ -123,6 +123,53 @@ Retrieve the metadata parsed from a data file.
...
+
+    label = meta.filelabel
+    if label != ""
+        ret["label"] = label
+    end
+
+    return ret
+end
+
+function metadata(tb::ReadStatTable, col::Symbol)
```
Comment (bkamins) [inline src/table.jl:144]: this should be `colmetadata`

**Comment by bkamins** [comment]
@nalimilan - do you want me to update this PR after DataAPI.jl 1.13 is out, or you want to do it (to practice the API 😄)?

**Comment by junyuan-chen** [comment]
> @nalimilan - do you want me to update this PR after DataAPI.jl 1.13 is out, or you want to do it (to practice the API 😄)?

@bkamins Thanks for looking into that! But, I forgot to mention that the one I pushed yesterday is still the old one written back in May. I have started rewriting it for the `DataAPI.jl` v1.13 and turned this PR to draft.

**Comment by junyuan-chen** [comment]
My attempt on the rewrite is now complete (except for documentation pages).

Here is a summary of main changes:

1. A `ReadStatTable` always contains metadata with a fixed set of keys and value types that are processed by `ReadStat`.
2. `ReadStatMeta` is split into `ReadStatMeta` and `ReadStatColMeta` for file-level and variable-level metadata respectively. They are the types for objects returned by `metadata` and `colmetadata` and can be iterated as key-value pairs.
3.  `ReadStatColMeta` is immutable. However, with `StructVector`, the metadata for each column can still be modified in-place as the underlying data are still stored in arrays.
4. Considering the changes to be included in v1.13 of `DataAPI.jl`, additional iterators `MetaStyleView` and `ColMetaIterator` are defined for lazy behavior that does not require allocating a `Dict`. The former is for the case when metadata style is needed; the latter is for iterating metadata by column.
5. Additionally, values of column-specific metadata for a specific key across different columns can be iterated using the array returned by `colmetavalues`.

**Comment by bkamins** [comment]
This is what RData.jl also does - right?

**Comment by junyuan-chen** [comment]
> It would make sense to mark labels with the `:note` style by default IMO, as these are the typical usage for `:note`.

@nalimilan @bkamins Sure. Let me do this change for the next release.

**Pull request merged**